receiver/oktareceiver/logs.go

```diff
@@ -0,0 +1,237 @@
+}
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// newOktaLogsReceiver returns a newly configured oktaLogsReceiver
+func newOktaLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) (*oktaLogsReceiver, error) {
+	var startTime time.Time
+	if cfg.StartTime != "" {
+		cfgStartTime, err := time.Parse(OktaTimeFormat, cfg.StartTime)
+		if err != nil {
+			return nil, err
+		}
+		startTime = cfgStartTime
+	} else {
+		startTime = time.Now().UTC().Add(-cfg.PollInterval)
+	}
+
+	return &oktaLogsReceiver{
+		cfg:       *cfg,
+		client:    http.DefaultClient,
+		consumer:  consumer,
+		doneChan:  make(chan bool),
+		logger:    logger,
+		startTime: startTime,
+		wg:        &sync.WaitGroup{},
+	}, nil
+}
+
+func (r *oktaLogsReceiver) Start(ctx context.Context, _ component.Host) error {
+	r.wg.Add(1)
+	go r.startPolling(ctx)
+	return nil
+}
+
+func (r *oktaLogsReceiver) startPolling(ctx context.Context) {
+	defer r.wg.Done()
+	t := time.NewTicker(r.cfg.PollInterval)
+
+	err := r.poll(ctx)
+	if err != nil {
+		r.logger.Error("there was an error during the first poll", zap.Error(err))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-r.doneChan:
+			return
+		case <-t.C:
+			err := r.poll(ctx)
+			if err != nil {
+				r.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *oktaLogsReceiver) poll(ctx context.Context) error {
+	select {
+	case _, ok := <-r.doneChan:
+		if !ok {
+			return nil
+		}
+	default:
+		logEvents := r.getLogs()
+		observedTime := pcommon.NewTimestampFromTime(time.Now())
+		logs := r.processLogEvents(observedTime, logEvents)
+		if logs.LogRecordCount() > 0 {
+			if err := r.consumer.ConsumeLogs(ctx, logs); err != nil {
+				r.logger.Error("unable to consume logs", zap.Error(err))
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (r *oktaLogsReceiver) getLogs() []*okta.LogEvent {
+	var req *http.Request
+	var err error
+	var logs []*okta.LogEvent
+	if r.nextURL == "" {
+		// for the first polling request, use startTime
+		reqURL := "https://" + r.cfg.Domain + "/api/v1/logs"
+		req, err = http.NewRequest("GET", reqURL, nil)
+		if err != nil {
+			r.logger.Warn("error creating okta api request", zap.Error(err))
+			return logs
+		}
+
+		query := req.URL.Query()
+		query.Add("since", r.startTime.Format(OktaTimeFormat))
+		req.URL.RawQuery = query.Encode()
+	} else {
+		req, err = http.NewRequest("GET", r.nextURL, nil)
+		if err != nil {
+			r.logger.Warn("error creating okta api request", zap.Error(err))
+			return logs
+		}
+	}
+
+	req.Header.Add("Authorization", "SSWS "+r.cfg.APIToken)
+	res, err := r.client.Do(req)
+	if err != nil {
+		r.logger.Warn("error performing okta api request", zap.Error(err))
+		return logs
+	}
+
+	defer res.Body.Close()
+
+	if res.StatusCode != 200 {
+		r.logger.Error("okta logs endpoint returned non-200 statuscode: " + res.Status)
+		return logs
+	}
+
+	r.setNextLink(res)
```

Comment:
Great catch here as well,
I just pushed up some changes that fix this case where there are more than 1000 logs happening between polling intervals. I also added a limit parameter to the API query and set it to 1000 in order to minimize these edge cases happening. It was previously unset and the default is 100.

And actually we don't have to do 2 requests for most polls, we can just check if the # of logs returned is less than 1000 which means there are no "overflow" logs. And if there are 1000, we check if the 1000th log has a published timestamp after our initial Poll timestamp.

Another thing to note:
Okta API also mentions that with polling type requests, the logs are not _guaranteed_ to be in strictly ascending order based on published timestamp, however from my local testing hitting the okta API, I can see that they are still at least close to ascending order so I think this solution will work just fine.

There is an edge case where okta returns 1000 logs, and then 999th log has a published timestamp after our initial poll, but the 1000th log is out of order and has a timestamp before our initial poll, which in that case with the logic I just pushed, it would make another request to the log endpoint to get the next "page". 

Otherwise to make sure they are in strictly ascending order we could sort all the logs that come back if there are 1000 logs returned which I don't think would be worth the extra complexity.